eth_worker/sql_persistence/interface.py

```diff
@@ -444,6 +471,43 @@ def set_wallet_last_topup_task_uuid(self, wallet_address, task_uuid):
 
         self.session.commit()
 
+    def check_if_synchronization_filter_exists(self, contract_address, filter_parameters):
+        filter = self.session.query(SynchronizationFilter).filter(
+            SynchronizationFilter.contract_address == contract_address, SynchronizationFilter.filter_parameters == filter_parameters).first()
+        if filter == None:
+            return False
+        return filter
+
+    def add_transaction_filter(self, contract_address, contract_type, filter_parameters, filter_type, decimals, block_epoch):
+        filter = SynchronizationFilter(contract_address=contract_address, contract_type=contract_type, filter_parameters=filter_parameters, max_block=block_epoch, filter_type=filter_type, decimals=decimals)
+        self.session.add(filter)
+        self.session.commit()
+        return filter
+
+    def set_filter_max_block(self, filter_id, block):
+        filter = self.session.query(SynchronizationFilter).filter(SynchronizationFilter.id == filter_id).first()
+        filter.max_block = block
+        self.session.commit()
+
+    def get_synchronization_filter(self, filter_id):
+        return self.session.query(SynchronizationFilter).filter(SynchronizationFilter.id == filter_id).first()
+
+    def get_all_synchronization_filters(self):
+        return self.session.query(SynchronizationFilter).all()
+
+    def add_block_range(self, start, end, filter_id):
+        for n in range(start, end):
+            block = SynchronizedBlock(block_number = n , status = 'PENDING', is_synchronized = False, synchronization_filter_id = filter_id)
+            self.session.add(block)
+        self.session.commit()
+
+    def set_block_range_status(self, start, end, status):
+        for n in range(start, end):
+            block  = self.session.query(SynchronizedBlock).filter(SynchronizedBlock.block_number == n).first()
```

Comment:
`.first()` is problematic since it's possible for the same block to be added more than once. (assumption of non-concurrency above)

Reply:
Changed to all(), but also added a lock so there should never be multiple now either